File: Creational/AbstractFactory/Java/Structure/ConcreteProductB2.java
/**
 * Concrete implementation of ProductB for family 2
 */
public class ConcreteProductB2 implements ProductB {
    
    @Override
    public void operationB() {
        System.out.println("ConcreteProductB2: Operation B executed");
    }
    
    @Override
    public void collaborateWithA(ProductA productA) {
        System.out.println("ConcreteProductB2: Collaborating with " + productA.getClass().getSimpleName());
    }
}

File: Creational/AbstractFactory/TypeScript/Structure/ConcreteProductB2.ts
import { AbstractProductB } from "./AbstractProductB";
import { AbstractProductA } from "./AbstractProductA";

/**
 * ConcreteProductB2 - Concrete implementation of AbstractProductB
 * Represents Product B from Family 2
 */
export class ConcreteProductB2 extends AbstractProductB {
  operationB(): string {
    return "ConcreteProductB2 operation";
  }

  getFamily(): string {
    return "Family 2";
  }

  collaborateWith(productA: AbstractProductA): string {
    return `ConcreteProductB2 collaborating with ${productA.constructor.name}`;
  }
}
